In [1]:
import pandas as pd
import numpy as np

In [2]:
df= pd.read_csv('ipl.csv')
pd.set_option('display.max_columns', None)
df.head(50)

C:\Users\admin\AppData\Local\Temp\ipykernel_8376\896356783.py:1: DtypeWarning: Columns (1,11) have mixed types. Specify dtype option on import or set low_memory=False.
  df= pd.read_csv('ipl.csv')


,match_id,season,date,city,venue,team1,team2,toss_winner,toss_decision,winner,result,result_margin,player_of_match,inning,batting_team,bowling_team,over,ball,batter,bowler,non_striker,runs_scored,extras,current_score,wickets_down,balls_remaining,wickets_remaining,current_run_rate,required_run_rate,target_score,wicket_kind,player_out,fielder
0,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,1,SC Ganguly,P Kumar,BB McCullum,0,1,1,0,119,10,6.00,NaN,NaN,NaN,NaN,NaN
1,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,2,BB McCullum,P Kumar,SC Ganguly,0,0,1,0,118,10,3.00,NaN,NaN,NaN,NaN,NaN
2,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,BB McCullum,P Kumar,SC Ganguly,0,1,2,0,118,10,6.00,NaN,NaN,NaN,NaN,NaN
3,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,4,BB McCullum,P Kumar,SC Ganguly,0,0,2,0,117,10,4.00,NaN,NaN,NaN,NaN,NaN
4,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,5,BB McCullum,P Kumar,SC Ganguly,0,0,2,0,116,10,3.00,NaN,NaN,NaN,NaN,NaN
5,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,6,BB McCullum,P Kumar,SC Ganguly,0,0,2,0,115,10,2.40,NaN,NaN,NaN,NaN,NaN
6,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,7,BB McCullum,P Kumar,SC Ganguly,0,1,3,0,114,10,3.00,NaN,NaN,NaN,NaN,NaN
7,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,1,1,BB McCullum,Z Khan,SC Ganguly,0,0,3,0,113,10,2.57,NaN,NaN,NaN,NaN,NaN
8,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,1,2,BB McCullum,Z Khan,SC Ganguly,4,0,7,0,112,10,5.25,NaN,NaN,NaN,NaN,NaN
9,335982,2007/08,4/18/2008,Bangalore,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140,BB McCullum,1,Kolkata Knight Riders,Royal Challengers Bangalore,1,3,BB McCullum,Z Khan,SC Ganguly,4,0,11,0,111,10,7.33,NaN,NaN,NaN,NaN,NaN


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278205 entries, 0 to 278204
Data columns (total 33 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   match_id           278205 non-null  int64  
 1   season             278205 non-null  object 
 2   date               278205 non-null  object 
 3   city               265808 non-null  object 
 4   venue              278205 non-null  object 
 5   team1              278205 non-null  object 
 6   team2              278205 non-null  object 
 7   toss_winner        278205 non-null  object 
 8   toss_decision      278205 non-null  object 
 9   winner             273503 non-null  object 
 10  result             278205 non-null  object 
 11  result_margin      278205 non-null  object 
 12  player_of_match    277399 non-null  object 
 13  inning             278205 non-null  int64  
 14  batting_team       278205 non-null  object 
 15  bowling_team       278205 non-null  object 
 16  ov

In [4]:
for col in df.select_dtypes(include='object').columns:
    print(f"\n--- {col} ---")
    print(df[col].value_counts())


--- season ---
season
2013       18177
2022       17912
2023       17863
2012       17767
2025       17285
2024       17103
2020/21    14559
2009/10    14498
2019       14312
2014       14300
2016       14096
2017       13862
2015       13652
2009       13606
2007/08    13489
2018       13380
2021       11434
2011        9454
2011        7559
2021        2991
2018         906
Name: count, dtype: int64

--- date ---
date
10/18/2020    520
4/8/2025      517
4/23/2009     513
4/30/2023     510
4/25/2021     510
             ... 
5/13/2015     126
5/5/2025      123
5/15/2015     106
5/8/2025       63
4/30/2019      51
Name: count, Length: 885, dtype: int64

--- city ---
city
Mumbai            43374
Kolkata           23436
Delhi             22993
Chennai           22132
Hyderabad         19664
Jaipur            15194
Bangalore         14907
Chandigarh        14476
Pune              12239
Ahmedabad         10816
Abu Dhabi          8799
Bengaluru          8043
Lucknow            5235
Visakha

In [5]:
cols = ['required_run_rate', 'target_score', 'wicket_kind', 'player_out', 'fielder']

for col in cols:
    print(f"\n{'='*50}")
    print(f"{col}")
    print(df[col].value_counts(dropna=False))


required_run_rate
required_run_rate
 NaN      144302
 6.00       1294
 12.00      1158
 9.00       1076
 8.00        609
           ...  
-1.29          1
 18.47         1
 22.53         1
 21.22         1
 20.38         1
Name: count, Length: 3075, dtype: int64

target_score
target_score
NaN      144302
172.0      2548
163.0      2519
166.0      2470
165.0      2447
          ...  
48.0         35
66.0         34
54.0         32
43.0         30
63.0         20
Name: count, Length: 176, dtype: int64

wicket_kind
wicket_kind
NaN                      264382
caught                     8665
bowled                     2345
run out                    1153
lbw                         853
caught and bowled           388
stumped                     376
hit wicket                   18
retired hurt                 17
retired out                   5
obstructing the field         3
Name: count, dtype: int64

player_out
player_out
NaN             264382
RG Sharma          237
V Kohli            219

### Missing Value Treatment

Missing values were analyzed based on the meaning of each variable.

- Null values in **`wicket_kind`**, **`player_out`**, and **`fielder`** were retained because they represent deliveries where no dismissal occurred.
- Similarly, **`target_score`** and **`required_run_rate`** were retained as missing where these metrics were not applicable.

Therefore, these null values were **not removed or replaced**, as they represent valid and meaningful missing values in the cricket dataset.

In [6]:
missing = df.isnull().sum().sort_values(ascending=False)

print(missing)

fielder              268192
player_out           264382
wicket_kind          264382
target_score         144302
required_run_rate    144302
city                  12397
winner                 4702
player_of_match         806
match_id                  0
toss_decision             0
toss_winner               0
result                    0
season                    0
team1                     0
team2                     0
date                      0
venue                     0
over                      0
bowling_team              0
batting_team              0
inning                    0
result_margin             0
ball                      0
bowler                    0
batter                    0
wickets_down              0
current_score             0
extras                    0
runs_scored               0
non_striker               0
current_run_rate          0
balls_remaining           0
wickets_remaining         0
dtype: int64


In [7]:
for col in df.columns:
    print(f"\n{'=' * 60}")
    print(f"        {col.upper()}")
    print(f"{'=' * 60}")
    print("Unique values:", df[col].unique())
    print("Missing values:", df[col].isnull().sum())


        MATCH_ID
Unique values: [ 335982  335983  335984 ... 1473510 1473511 1485779]
Missing values: 0

        SEASON
Unique values: ['2007/08' '2009' '2009/10' '2011' 2011 2012 2013 2014 2015 2016 2017 2018
 '2018' '2019' '2020/21' '2021' 2021 2022 2023 2024 2025]
Missing values: 0

        DATE
Unique values: ['4/18/2008' '4/19/2008' '4/20/2008' '4/21/2008' '4/22/2008' '4/23/2008'
 '4/24/2008' '4/25/2008' '4/26/2008' '4/27/2008' '4/28/2008' '4/29/2008'
 '4/30/2008' '5/1/2008' '5/2/2008' '5/25/2008' '5/3/2008' '5/4/2008'
 '5/5/2008' '5/6/2008' '5/7/2008' '5/8/2008' '5/9/2008' '5/28/2008'
 '5/10/2008' '5/11/2008' '5/12/2008' '5/13/2008' '5/14/2008' '5/15/2008'
 '5/16/2008' '5/17/2008' '5/18/2008' '5/19/2008' '5/20/2008' '5/21/2008'
 '5/23/2008' '5/24/2008' '5/26/2008' '5/27/2008' '5/30/2008' '5/31/2008'
 '6/1/2008' '4/18/2009' '4/19/2009' '4/20/2009' '4/21/2009' '4/22/2009'
 '4/23/2009' '4/24/2009' '4/25/2009' '4/26/2009' '4/27/2009' '4/28/2009'
 '4/29/2009' '4/30/2009' '5/1/2009' '

In [8]:
missing_summary = pd.DataFrame({
    'Column': df.columns,
    'Missing Values': df.isnull().sum(),
    'Missing Percentage': (df.isnull().sum() / len(df) * 100).round(2)
})

missing_summary

,Column,Missing Values,Missing Percentage
match_id,match_id,0,0.00
season,season,0,0.00
date,date,0,0.00
city,city,12397,4.46
venue,venue,0,0.00
team1,team1,0,0.00
team2,team2,0,0.00
toss_winner,toss_winner,0,0.00
toss_decision,toss_decision,0,0.00
winner,winner,4702,1.69


### Missing Value Analysis

The columns **`wicket_kind`**, **`player_out`**, and **`fielder`** contain a high percentage of missing values (95.03%, 95.03%, and 96.40%, respectively). These missing values were retained because they are expected in deliveries where no relevant dismissal event occurred. Therefore, they were not treated as erroneous missing data.

In [9]:
df[df['city'].isnull()]['venue'].value_counts()

venue
Dubai International Cricket Stadium    8080
Sharjah Cricket Stadium                4317
Name: count, dtype: int64

In [10]:
## Treating City column missing values
df.loc[
    (df['city'].isnull()) & 
    (df['venue'] == 'Dubai International Cricket Stadium'),
    'city'
] = 'Dubai'

df.loc[
    (df['city'].isnull()) & 
    (df['venue'] == 'Sharjah Cricket Stadium'),
    'city'
] = 'Sharjah'

In [11]:
df['city'].isnull().sum()

np.int64(0)

### Winner column

In [12]:
df[df['winner'].isnull()]['result'].value_counts(dropna=False)

result
wickets    4702
Name: count, dtype: int64

In [13]:
df.columns.tolist()

['match_id',
 'season',
 'date',
 'city',
 'venue',
 'team1',
 'team2',
 'toss_winner',
 'toss_decision',
 'winner',
 'result',
 'result_margin',
 'player_of_match',
 'inning',
 'batting_team',
 'bowling_team',
 'over',
 'ball',
 'batter',
 'bowler',
 'non_striker',
 'runs_scored',
 'extras',
 'current_score',
 'wickets_down',
 'balls_remaining',
 'wickets_remaining',
 'current_run_rate',
 'required_run_rate',
 'target_score',
 'wicket_kind',
 'player_out',
 'fielder']

In [14]:
df[df['winner'].isnull()][['match_id', 'team1', 'team2', 'result']].drop_duplicates()

,match_id,team1,team2,result
15161,392190,Kolkata Knight Riders,Rajasthan Royals,wickets
30707,419121,Chennai Super Kings,Kings XI Punjab,wickets
57068,501265,Delhi Daredevils,Pune Warriors,wickets
77794,598004,Sunrisers Hyderabad,Royal Challengers Bangalore,wickets
80965,598017,Royal Challengers Bangalore,Delhi Daredevils,wickets
98800,729315,Kolkata Knight Riders,Rajasthan Royals,wickets
113250,829741,Rajasthan Royals,Kings XI Punjab,wickets
115599,829763,Royal Challengers Bangalore,Rajasthan Royals,wickets
121177,829813,Royal Challengers Bangalore,Delhi Daredevils,wickets
144440,1082625,Gujarat Lions,Mumbai Indians,wickets


In [15]:
df[df['winner'].isnull()][
    ['match_id', 'inning', 'batting_team', 'result']
].drop_duplicates().sort_values(['match_id', 'inning'])

,match_id,inning,batting_team,result
15161,392190,1,Rajasthan Royals,wickets
15287,392190,2,Kolkata Knight Riders,wickets
15417,392190,3,Kolkata Knight Riders,wickets
15424,392190,4,Rajasthan Royals,wickets
30707,419121,1,Kings XI Punjab,wickets
...,...,...,...,...
268564,1473469,4,Delhi Capitals,wickets
271126,1473481,1,Punjab Kings,wickets
271249,1473481,2,Kolkata Knight Riders,wickets
273705,1473492,1,Delhi Capitals,wickets


In [16]:
last_inning_team = (
    df.groupby('match_id')
      .apply(lambda x: x.loc[x['inning'].idxmax(), 'batting_team'])
)

known_winner = (
    df[['match_id', 'winner']]
      .dropna(subset=['winner'])
      .drop_duplicates('match_id')
)

check = known_winner.copy()

check['last_inning_team'] = check['match_id'].map(last_inning_team)

check['matches'] = check['winner'] == check['last_inning_team']

check['matches'].value_counts()

C:\Users\admin\AppData\Local\Temp\ipykernel_8376\3374671757.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: x.loc[x['inning'].idxmax(), 'batting_team'])


matches
True     621
False    525
Name: count, dtype: int64

In [17]:
match_info = df[['match_id', 'winner', 'result']].drop_duplicates('match_id')

second_inning_team = (
    df[df['inning'] == 2]
    .groupby('match_id')['batting_team']
    .first()
)

check = match_info[
    (match_info['result'] == 'wickets') &
    (match_info['winner'].notna())
].copy()

check['second_inning_team'] = check['match_id'].map(second_inning_team)

check['winner_matches'] = (
    check['winner'] == check['second_inning_team']
)

check['winner_matches'].value_counts()

winner_matches
True    615
Name: count, dtype: int64

In [18]:
second_inning_team = (
    df[df['inning'] == 2]
    .groupby('match_id')['batting_team']
    .first()
)

df['winner'] = df['winner'].fillna(
    df['match_id'].map(second_inning_team)
)

In [19]:
df['winner'].isnull().sum()

np.int64(495)

In [20]:
remaining_winner = df[df['winner'].isnull()][
    ['match_id', 'inning', 'batting_team', 'team1', 'team2', 'result']
].drop_duplicates().sort_values(['match_id', 'inning'])

remaining_winner

,match_id,inning,batting_team,team1,team2,result
57068,501265,1,Delhi Daredevils,Delhi Daredevils,Pune Warriors,wickets
115599,829763,1,Royal Challengers Bangalore,Royal Challengers Bangalore,Rajasthan Royals,wickets
236750,1359519,1,Lucknow Super Giants,Lucknow Super Giants,Chennai Super Kings,wickets
273705,1473492,1,Delhi Capitals,Delhi Capitals,Sunrisers Hyderabad,wickets
274316,1473495,1,Punjab Kings,Punjab Kings,Delhi Capitals,wickets


In [21]:
df[df['winner'].isnull()]['inning'].value_counts()

inning
1    495
Name: count, dtype: int64

In [22]:
# 1. Check remaining missing values
print("Missing winner values:", df['winner'].isnull().sum())

# 2. Check that each match has at most one unique winner
winner_per_match = df.groupby('match_id')['winner'].nunique(dropna=True)
print("Maximum unique winners in a match:", winner_per_match.max())

# 3. Check whether winner is one of the two teams
valid_winner = (
    df['winner'].isna() |
    (df['winner'] == df['team1']) |
    (df['winner'] == df['team2'])
)

print("Invalid winner records:", (~valid_winner).sum())

# 4. Check filled winners against second innings
second_inning_team = (
    df[df['inning'] == 2]
    .groupby('match_id')['batting_team']
    .first()
)

wicket_matches = df[
    (df['result'] == 'wickets') &
    (df['winner'].notna())
].drop_duplicates('match_id').copy()

wicket_matches['second_inning_team'] = (
    wicket_matches['match_id'].map(second_inning_team)
)

wicket_matches['verified'] = (
    wicket_matches['winner'] == wicket_matches['second_inning_team']
)

print("Wicket-result matches verified:")
print(wicket_matches['verified'].value_counts(dropna=False))

Missing winner values: 495
Maximum unique winners in a match: 1
Invalid winner records: 0
Wicket-result matches verified:
verified
True    633
Name: count, dtype: int64


### Winner Column – Missing Value Treatment

The **`winner`** column was analyzed using **`match_id`**, **`inning`**, and **`batting_team`**.

Missing winner values were filled where the second-innings batting team could reliably determine the match winner. This resulted in **4,207 values being filled**.

A total of **495 values** remained missing because only first-innings information was available, making it impossible to determine the winner reliably. These values were retained as `NaN` rather than making assumptions.

The cleaned column was verified to ensure that each match has at most one winner and that all available winners belong to either **`team1`** or **`team2`**.

### Player of Match

In [23]:
df[df['player_of_match'].isnull()]['result'].value_counts(dropna=False)

result
wickets    806
Name: count, dtype: int64

In [24]:
pom_check = df.groupby('match_id')['player_of_match'].nunique(dropna=True)

print("Matches with player_of_match available:", (pom_check > 0).sum())
print("Matches with player_of_match completely missing:", (pom_check == 0).sum())
print("Maximum unique player_of_match per match:", pom_check.max())

Matches with player_of_match available: 1161
Matches with player_of_match completely missing: 8
Maximum unique player_of_match per match: 1


In [25]:
df[df['player_of_match'].isnull()][
    ['match_id', 'team1', 'team2', 'result']
].drop_duplicates()

,match_id,team1,team2,result
57068,501265,Delhi Daredevils,Pune Warriors,wickets
115599,829763,Royal Challengers Bangalore,Rajasthan Royals,wickets
121177,829813,Royal Challengers Bangalore,Delhi Daredevils,wickets
176392,1178424,Royal Challengers Bangalore,Rajasthan Royals,wickets
236750,1359519,Lucknow Super Giants,Chennai Super Kings,wickets
271126,1473481,Punjab Kings,Kolkata Knight Riders,wickets
273705,1473492,Delhi Capitals,Sunrisers Hyderabad,wickets
274316,1473495,Punjab Kings,Delhi Capitals,wickets


In [26]:
pom_missing_matches = df[df['player_of_match'].isnull()]['match_id'].unique()

df[df['match_id'].isin(pom_missing_matches)][
    ['match_id', 'player_of_match']
].drop_duplicates()

,match_id,player_of_match
57068,501265,NaN
115599,829763,NaN
121177,829813,NaN
176392,1178424,NaN
236750,1359519,NaN
271126,1473481,NaN
273705,1473492,NaN
274316,1473495,NaN


In [27]:
print("Remaining missing values:", df['player_of_match'].isnull().sum())
print("Unique players per match:", df.groupby('match_id')['player_of_match'].nunique(dropna=True).max())

Remaining missing values: 806
Unique players per match: 1


### Player of the Match – Missing Value Treatment

The **`player_of_match`** column contained **806 missing values**. The missing values were checked at the **`match_id`** level to determine whether the player name was available in another row of the same match.

The analysis showed that these missing values belonged to matches where **`player_of_match`** was missing throughout the entire match. Therefore, the missing values could not be reliably recovered from the available dataset.

Hence, these values were **retained as `NaN`** rather than using arbitrary or assumed values.

## Result margin

In [28]:
print("Data type:", df['result_margin'].dtype)

print("\nUnique values:")
print(df['result_margin'].unique())

print("\nValue counts:")
print(df['result_margin'].value_counts(dropna=False))

print("\nStatistics:")
print(df['result_margin'].describe())


Data type: object

Unique values:
['140' '33' '9' '5' '6' '3' '66' '7' '10' '4' '13' '45' '8' '29' '18' '23'
 '41' '12' '65' '25' '1' '14' '105' '19' '75' '92' '11' '24' 'undefined'
 '27' '38' '78' '16' '53' '2' '31' '55' '98' '34' '36' '39' '17' 17 41 40
 4 5 67 24 6 23 63 8 37 2 9 14 7 13 39 57 11 35 38 22 33 21 3 16 48 29 25
 55 '26' '20' '21' '85' '32' '63' '76' '111' '82' '43' '58' '28' '74' '22'
 '42' '59' '46' '37' '47' '35' '86' '44' '48' '87' 86 130 58 15 60 65 77
 46 30 50 93 72 '15' '62' '30' '97' '138' '71' '144' '80' '51' '61' '146'
 10 20 1 19 71 64 31 102 34 '118' '40' '49' '57' '88' '60' '69' '54' 12 18
 36 75 91 52 62 54 81 '56' '112' '77' '81' 28 106 78 98 47 27 44 '50'
 '100' '83' '110']

Value counts:
result_margin
6      25255
7      21391
5      19756
8      14548
4      12945
       ...  
75       211
146      211
76       204
112      189
86        17
Name: count, Length: 161, dtype: int64

Statistics:
count     278205
unique       161
top            6
freq     

In [29]:
df[df['result_margin'] == 'undefined']['result'].value_counts(dropna=False)

result
wickets    4702
Name: count, dtype: int64

In [30]:
print("Undefined values:", (df['result_margin'] == 'undefined').sum())

Undefined values: 4702


In [31]:
wicket_matches = df[df['result'] == 'wickets']

wicket_matches.groupby('match_id').agg(
    result_margin=('result_margin', 'first'),
    final_wickets_remaining=('wickets_remaining', 'last')
).head(20)

,result_margin,final_wickets_remaining
match_id,,
335984,9,9
335985,5,5
335986,5,5
335987,6,6
335988,9,9
335990,3,3
335992,7,7
335993,9,9
335994,10,10


In [32]:
undefined_check = df[df['result_margin'] == 'undefined'].groupby('match_id').agg(
    result=('result', 'first'),
    final_wickets_remaining=('wickets_remaining', 'last')
)

undefined_check.head(20)

,result,final_wickets_remaining
match_id,,
392190,wickets,10
419121,wickets,9
501265,wickets,7
598004,wickets,10
598017,wickets,8
729315,wickets,10
829741,wickets,8
829763,wickets,3
829813,wickets,10


In [33]:
df.drop('result_margin', axis=1, inplace=True)

In [34]:
print('result_margin' in df.columns)

False


### Removing `result_margin`

The **`result_margin`** column was removed because it contained inconsistent data types and **4,702 `undefined` values**. The missing values could not be reliably recovered from the available records. Since this column was not essential for the planned analysis and dashboard, it was removed to maintain data quality and consistency.

## Date

In [35]:
print("Data type:", df['date'].dtype)

print("\nMissing values:", df['date'].isnull().sum())

print("\nSample values:")
print(df['date'].head(10))

print("\nUnique values:", df['date'].nunique())

Data type: object

Missing values: 0

Sample values:
0    4/18/2008
1    4/18/2008
2    4/18/2008
3    4/18/2008
4    4/18/2008
5    4/18/2008
6    4/18/2008
7    4/18/2008
8    4/18/2008
9    4/18/2008
Name: date, dtype: object

Unique values: 885


In [36]:
df.isnull().sum().sort_values(ascending=False)

fielder              268192
wicket_kind          264382
player_out           264382
required_run_rate    144302
target_score         144302
player_of_match         806
winner                  495
match_id                  0
toss_winner               0
team2                     0
toss_decision             0
season                    0
venue                     0
team1                     0
date                      0
city                      0
over                      0
bowling_team              0
batting_team              0
inning                    0
result                    0
ball                      0
bowler                    0
batter                    0
wickets_down              0
current_score             0
extras                    0
runs_scored               0
non_striker               0
current_run_rate          0
balls_remaining           0
wickets_remaining         0
dtype: int64

In [37]:
print(df[df['required_run_rate'].isnull()]['inning'].value_counts())

inning
1    144131
3        83
4        76
5         8
6         4
Name: count, dtype: int64


In [38]:
print(df[df['target_score'].isnull()]['inning'].value_counts())

inning
1    144131
3        83
4        76
5         8
6         4
Name: count, dtype: int64


In [39]:
print(
    df[df['required_run_rate'].isnull()]
    .groupby(['inning', 'result'])
    .size()
)

inning  result 
1       runs       65910
        wickets    78221
3       wickets       83
4       wickets       76
5       wickets        8
6       wickets        4
dtype: int64


In [40]:
print(
    df[df['target_score'].isnull()]
    .groupby(['inning', 'result'])
    .size()
)

inning  result 
1       runs       65910
        wickets    78221
3       wickets       83
4       wickets       76
5       wickets        8
6       wickets        4
dtype: int64


In [41]:
df[
    df['required_run_rate'].isnull() &
    (df['inning'] > 1)
][
    ['match_id', 'inning', 'batting_team', 'bowling_team',
     'result', 'current_score', 'target_score',
     'required_run_rate']
].drop_duplicates().head(50)

,match_id,inning,batting_team,bowling_team,result,current_score,target_score,required_run_rate
15417,392190,3,Kolkata Knight Riders,Rajasthan Royals,wickets,1,NaN,NaN
15418,392190,3,Kolkata Knight Riders,Rajasthan Royals,wickets,2,NaN,NaN
15419,392190,3,Kolkata Knight Riders,Rajasthan Royals,wickets,3,NaN,NaN
15420,392190,3,Kolkata Knight Riders,Rajasthan Royals,wickets,7,NaN,NaN
15421,392190,3,Kolkata Knight Riders,Rajasthan Royals,wickets,11,NaN,NaN
15422,392190,3,Kolkata Knight Riders,Rajasthan Royals,wickets,15,NaN,NaN
15424,392190,4,Rajasthan Royals,Kolkata Knight Riders,wickets,6,NaN,NaN
15425,392190,4,Rajasthan Royals,Kolkata Knight Riders,wickets,8,NaN,NaN
15426,392190,4,Rajasthan Royals,Kolkata Knight Riders,wickets,14,NaN,NaN
15427,392190,4,Rajasthan Royals,Kolkata Knight Riders,wickets,18,NaN,NaN


### Missing Value Treatment – Target Score and Required Run Rate

The **`target_score`** and **`required_run_rate`** columns contain missing values primarily in the first innings, where these variables are not applicable because no target has been set yet.

A small number of missing values were also observed in later innings (innings 3–6), all associated with wicket-based results and special match situations. These values were retained as **NaN** because there was insufficient information to reliably derive them.

Therefore, no imputation was performed for **`target_score`** or **`required_run_rate`**.

In [42]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [43]:
## Teams name check
team_cols = [
    'team1',
    'team2',
    'toss_winner',
    'winner',
    'batting_team',
    'bowling_team'
]

for col in team_cols:
    print(f"\n{'=' * 50}")
    print(f"{col.upper()}")
    print(f"{'=' * 50}")
    print(df[col].unique())


TEAM1
['Royal Challengers Bangalore' 'Kings XI Punjab' 'Delhi Daredevils'
 'Mumbai Indians' 'Kolkata Knight Riders' 'Rajasthan Royals'
 'Deccan Chargers' 'Chennai Super Kings' 'Kochi Tuskers Kerala'
 'Pune Warriors' 'Sunrisers Hyderabad' 'Gujarat Lions'
 'Rising Pune Supergiants' 'Rising Pune Supergiant' 'Delhi Capitals'
 'Punjab Kings' 'Lucknow Super Giants' 'Gujarat Titans'
 'Royal Challengers Bengaluru']

TEAM2
['Kolkata Knight Riders' 'Chennai Super Kings' 'Rajasthan Royals'
 'Royal Challengers Bangalore' 'Deccan Chargers' 'Kings XI Punjab'
 'Delhi Daredevils' 'Mumbai Indians' 'Kochi Tuskers Kerala'
 'Pune Warriors' 'Sunrisers Hyderabad' 'Rising Pune Supergiants'
 'Gujarat Lions' 'Rising Pune Supergiant' 'Delhi Capitals' 'Punjab Kings'
 'Gujarat Titans' 'Lucknow Super Giants' 'Royal Challengers Bengaluru']

TOSS_WINNER
['Royal Challengers Bangalore' 'Chennai Super Kings' 'Rajasthan Royals'
 'Mumbai Indians' 'Deccan Chargers' 'Kings XI Punjab'
 'Kolkata Knight Riders' 'Delhi Darede

In [44]:
print("Result values:")
print(df['result'].unique())

print("\nToss decision values:")
print(df['toss_decision'].unique())

Result values:
['runs' 'wickets']

Toss decision values:
['field' 'bat']


In [45]:
df.dtypes

match_id               int64
season                object
date                  object
city                  object
venue                 object
team1                 object
team2                 object
toss_winner           object
toss_decision         object
winner                object
result                object
player_of_match       object
inning                 int64
batting_team          object
bowling_team          object
over                   int64
ball                   int64
batter                object
bowler                object
non_striker           object
runs_scored            int64
extras                 int64
current_score          int64
wickets_down           int64
balls_remaining        int64
wickets_remaining      int64
current_run_rate     float64
required_run_rate    float64
target_score         float64
wicket_kind           object
player_out            object
fielder               object
dtype: object

In [46]:
num_cols = [
    'match_id',
    'season',
    'inning',
    'over',
    'ball',
    'runs_scored',
    'extras',
    'current_score',
    'wickets_down',
    'balls_remaining',
    'wickets_remaining',
    'current_run_rate',
    'required_run_rate',
    'target_score'
]

df[num_cols].dtypes

match_id               int64
season                object
inning                 int64
over                   int64
ball                   int64
runs_scored            int64
extras                 int64
current_score          int64
wickets_down           int64
balls_remaining        int64
wickets_remaining      int64
current_run_rate     float64
required_run_rate    float64
target_score         float64
dtype: object

In [47]:
print("Shape:", df.shape)

print("\nDuplicate rows:", df.duplicated().sum())

print("\nMissing values:")
print(df.isnull().sum().sort_values(ascending=False))

Shape: (278205, 32)

Duplicate rows: 0

Missing values:
fielder              268192
wicket_kind          264382
player_out           264382
required_run_rate    144302
target_score         144302
player_of_match         806
winner                  495
match_id                  0
toss_winner               0
team2                     0
toss_decision             0
season                    0
venue                     0
team1                     0
date                      0
city                      0
over                      0
bowling_team              0
batting_team              0
inning                    0
result                    0
ball                      0
bowler                    0
batter                    0
wickets_down              0
current_score             0
extras                    0
runs_scored               0
non_striker               0
current_run_rate          0
balls_remaining           0
wickets_remaining         0
dtype: int64


In [48]:
print("VENUE LIST")
print("=" * 60)

for i, venue in enumerate(sorted(df['venue'].dropna().unique()), 1):
    print(f"{i}. {venue}")

VENUE LIST
1. Arun Jaitley Stadium
2. Arun Jaitley Stadium, Delhi
3. Barabati Stadium
4. Barsapara Cricket Stadium, Guwahati
5. Bharat Ratna Shri Atal Bihari Vajpayee Ekana Cricket Stadium, Lucknow
6. Brabourne Stadium
7. Brabourne Stadium, Mumbai
8. Buffalo Park
9. De Beers Diamond Oval
10. Dr DY Patil Sports Academy
11. Dr DY Patil Sports Academy, Mumbai
12. Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium
13. Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium, Visakhapatnam
14. Dubai International Cricket Stadium
15. Eden Gardens
16. Eden Gardens, Kolkata
17. Feroz Shah Kotla
18. Green Park
19. Himachal Pradesh Cricket Association Stadium
20. Himachal Pradesh Cricket Association Stadium, Dharamsala
21. Holkar Cricket Stadium
22. JSCA International Stadium Complex
23. Kingsmead
24. M Chinnaswamy Stadium
25. M Chinnaswamy Stadium, Bengaluru
26. M.Chinnaswamy Stadium
27. MA Chidambaram Stadium
28. MA Chidambaram Stadium, Chepauk
29. MA Chidambaram Stadium, Chepauk, Chennai
30. Maha

In [49]:
venue_mapping = {

    # Delhi
    'Arun Jaitley Stadium': 'Arun Jaitley Stadium',
    'Arun Jaitley Stadium, Delhi': 'Arun Jaitley Stadium',
    'Feroz Shah Kotla': 'Arun Jaitley Stadium',

    # Brabourne
    'Brabourne Stadium': 'Brabourne Stadium',
    'Brabourne Stadium, Mumbai': 'Brabourne Stadium',

    # DY Patil
    'Dr DY Patil Sports Academy': 'Dr DY Patil Sports Academy',
    'Dr DY Patil Sports Academy, Mumbai': 'Dr DY Patil Sports Academy',

    # Visakhapatnam
    'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium':
        'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium',
    'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium, Visakhapatnam':
        'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium',

    # Eden Gardens
    'Eden Gardens': 'Eden Gardens',
    'Eden Gardens, Kolkata': 'Eden Gardens',

    # Himachal Pradesh
    'Himachal Pradesh Cricket Association Stadium':
        'Himachal Pradesh Cricket Association Stadium',
    'Himachal Pradesh Cricket Association Stadium, Dharamsala':
        'Himachal Pradesh Cricket Association Stadium',

    # Chinnaswamy
    'M Chinnaswamy Stadium': 'M Chinnaswamy Stadium',
    'M Chinnaswamy Stadium, Bengaluru': 'M Chinnaswamy Stadium',
    'M.Chinnaswamy Stadium': 'M Chinnaswamy Stadium',

    # Chepauk
    'MA Chidambaram Stadium': 'MA Chidambaram Stadium',
    'MA Chidambaram Stadium, Chepauk': 'MA Chidambaram Stadium',
    'MA Chidambaram Stadium, Chepauk, Chennai': 'MA Chidambaram Stadium',

    # Maharaja Yadavindra Singh Stadium
    'Maharaja Yadavindra Singh International Cricket Stadium, Mullanpur':
        'Maharaja Yadavindra Singh International Cricket Stadium',
    'Maharaja Yadavindra Singh International Cricket Stadium, New Chandigarh':
        'Maharaja Yadavindra Singh International Cricket Stadium',

    # Maharashtra Cricket Association
    'Maharashtra Cricket Association Stadium':
        'Maharashtra Cricket Association Stadium',
    'Maharashtra Cricket Association Stadium, Pune':
        'Maharashtra Cricket Association Stadium',
    'Subrata Roy Sahara Stadium':
        'Maharashtra Cricket Association Stadium',

    # Narendra Modi / Motera
    'Narendra Modi Stadium, Ahmedabad':
        'Narendra Modi Stadium',
    'Sardar Patel Stadium, Motera':
        'Narendra Modi Stadium',

    # Punjab Cricket Association
    'Punjab Cricket Association IS Bindra Stadium':
        'Punjab Cricket Association IS Bindra Stadium',
    'Punjab Cricket Association IS Bindra Stadium, Mohali':
        'Punjab Cricket Association IS Bindra Stadium',
    'Punjab Cricket Association IS Bindra Stadium, Mohali, Chandigarh':
        'Punjab Cricket Association IS Bindra Stadium',
    'Punjab Cricket Association Stadium, Mohali':
        'Punjab Cricket Association IS Bindra Stadium',

    # Rajiv Gandhi
    'Rajiv Gandhi International Stadium':
        'Rajiv Gandhi International Stadium',
    'Rajiv Gandhi International Stadium, Uppal':
        'Rajiv Gandhi International Stadium',
    'Rajiv Gandhi International Stadium, Uppal, Hyderabad':
        'Rajiv Gandhi International Stadium',

    # Sawai Mansingh
    'Sawai Mansingh Stadium': 'Sawai Mansingh Stadium',
    'Sawai Mansingh Stadium, Jaipur': 'Sawai Mansingh Stadium',

    # Wankhede
    'Wankhede Stadium': 'Wankhede Stadium',
    'Wankhede Stadium, Mumbai': 'Wankhede Stadium',

    # Abu Dhabi
    'Sheikh Zayed Stadium': 'Zayed Cricket Stadium, Abu Dhabi',
    'Zayed Cricket Stadium, Abu Dhabi': 'Zayed Cricket Stadium, Abu Dhabi'
}

df['venue'] = df['venue'].replace(venue_mapping)

In [50]:
print("Unique venues after cleaning:", df['venue'].nunique())

for i, venue in enumerate(sorted(df['venue'].dropna().unique()), 1):
    print(f"{i}. {venue}")

Unique venues after cleaning: 36
1. Arun Jaitley Stadium
2. Barabati Stadium
3. Barsapara Cricket Stadium, Guwahati
4. Bharat Ratna Shri Atal Bihari Vajpayee Ekana Cricket Stadium, Lucknow
5. Brabourne Stadium
6. Buffalo Park
7. De Beers Diamond Oval
8. Dr DY Patil Sports Academy
9. Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium
10. Dubai International Cricket Stadium
11. Eden Gardens
12. Green Park
13. Himachal Pradesh Cricket Association Stadium
14. Holkar Cricket Stadium
15. JSCA International Stadium Complex
16. Kingsmead
17. M Chinnaswamy Stadium
18. MA Chidambaram Stadium
19. Maharaja Yadavindra Singh International Cricket Stadium
20. Maharashtra Cricket Association Stadium
21. Narendra Modi Stadium
22. Nehru Stadium
23. New Wanderers Stadium
24. Newlands
25. OUTsurance Oval
26. Punjab Cricket Association IS Bindra Stadium
27. Rajiv Gandhi International Stadium
28. Saurashtra Cricket Association Stadium
29. Sawai Mansingh Stadium
30. Shaheed Veer Narayan Singh International 

In [ ]:
df.to_csv('ipl_cleaned_new.csv', index=False)

In [ ]:
print("Cleaned dataset saved successfully!")